cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -448,5 +626,455 @@ sample_edges(
                          std::move(labels));
 }
 
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          typename edge_type_t,
+          typename label_t>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<weight_t>>,
+           std::optional<rmm::device_uvector<edge_t>>,
+           std::optional<rmm::device_uvector<edge_type_t>>,
+           std::optional<rmm::device_uvector<int32_t>>,
+           std::optional<rmm::device_uvector<label_t>>,
+           std::optional<rmm::device_uvector<size_t>>>
+shuffle_and_organize_output(
+  raft::handle_t const& handle,
+  rmm::device_uvector<vertex_t>&& majors,
+  rmm::device_uvector<vertex_t>&& minors,
+  std::optional<rmm::device_uvector<weight_t>>&& weights,
+  std::optional<rmm::device_uvector<edge_t>>&& edge_ids,
+  std::optional<rmm::device_uvector<edge_type_t>>&& edge_types,
+  std::optional<rmm::device_uvector<int32_t>>&& hops,
+  std::optional<rmm::device_uvector<label_t>>&& labels,
+  std::optional<std::tuple<raft::device_span<label_t const>, raft::device_span<int32_t const>>>
+    label_to_output_comm_rank)
+{
+  std::optional<rmm::device_uvector<size_t>> offsets{std::nullopt};
+
+  if (label_to_output_comm_rank) {
+    CUGRAPH_EXPECTS(labels, "labels must be specified in order to shuffle sampling results");
+
+    auto& comm           = handle.get_comms();
+    auto const comm_size = comm.get_size();
+
+    auto total_global_mem = handle.get_device_properties().totalGlobalMem;
+    auto element_size     = sizeof(vertex_t) * 2 + (weights ? sizeof(weight_t) : size_t{0}) +
+                        (edge_ids ? sizeof(edge_t) : size_t{0}) +
+                        (edge_types ? sizeof(edge_type_t) : size_t{0}) +
+                        (hops ? sizeof(int32_t) : size_t{0}) + sizeof(label_t);
+
+    auto constexpr mem_frugal_ratio =
+      0.1;  // if the expected temporary buffer size exceeds the mem_frugal_ratio of the
+    // total_global_mem, switch to the memory frugal approach (thrust::sort is used to
+    // group-by by default, and thrust::sort requires temporary buffer comparable to the input
+    // data size)
+    auto mem_frugal_threshold =
+      static_cast<size_t>(static_cast<double>(total_global_mem / element_size) * mem_frugal_ratio);
+
+    // FIXME:  Need some MG tests that actually exercise all of these variations...
+
+    auto d_tx_value_counts = cugraph::groupby_and_count(
+      labels->begin(),
+      labels->end(),
+      [output_label = std::get<0>(*label_to_output_comm_rank),
+       output_rank  = std::get<1>(*label_to_output_comm_rank)] __device__(auto val) {
+        auto pos = thrust::find(thrust::seq, output_label.begin(), output_label.end(), val);
```

Comment:
`thrust::find` is linear scan, this can be expensive if output_label.size() is large. Better use binary search (if output_labels can be sorted) or a hash table. May look for prims/kv_store.cuh as well.

Reply:
I meant to go back and look at that, I will fix in next push.